components/raftstore/src/store/async_io/write_router.rs

```diff
@@ -238,39 +243,56 @@ where
                 let now = Instant::now();
                 if sender.send(msg, self.last_msg_priority).is_err() {
                     // Write threads are destroyed after store threads during shutdown.
-                    panic!("{} failed to send write msg, err: disconnected", self.tag);
+                    safe_panic!("{} failed to send write msg, err: disconnected", self.tag);
                 }
                 ctx.raft_metrics()
                     .write_block_wait
                     .observe(now.saturating_elapsed_secs());
             }
             Err(TrySendError::Disconnected(_)) => {
                 // Write threads are destroyed after store threads during shutdown.
-                panic!("{} failed to send write msg, err: disconnected", self.tag);
+                safe_panic!("{} failed to send write msg, err: disconnected", self.tag);
             }
         }
     }
 }
 
+pub type SenderVec<EK, ER> = Vec<Sender<WriteMsg<EK, ER>>>;
+
+#[derive(Clone)]
 /// Senders for asynchronous writes. There can be multiple senders, generally
 /// you should use `WriteRouter` to decide which sender to be used.
-#[derive(Clone)]
 pub struct WriteSenders<EK: KvEngine, ER: RaftEngine> {
-    write_senders: Vec<Sender<WriteMsg<EK, ER>>>,
+    senders: Tracker<SenderVec<EK, ER>>,
```

Comment:
How about `Tracker<Mutex<SenderVec<EK, ER>>>, then you don't need to impl Sync for Sender

Reply:
That would still make `last_msg_group` shared between threads. Is it okay, algorithm wise?

Reply:
Let's not. The implementation is wrong in one code path. And you can't staticly (or runtime for that matter) prevent concurrent access of that code path.

You can use Mutex instead of `VersionTrack`. Or if you prefer, you can refactor the implementation of `CachedTablet` to tikv_util.

Reply:
Or, now I think about it, you can make the inner `Vec<Sender<WriteMsg<EK, ER>>>` private, and only allow cloning senders (not allowing direct access).